Per Tenant Logging - Complete Implementation

shaangill025 opened the pull request:
resolve #2359

dbluhm commented:
I'm not terribly fond of this approach. In order to keep it going, we have to remember to get a logger instance with a profile which adds overhead to a task that should incur as little overhead as possible -- if not at execution time, it at least adds overhead in code. This means more lines of code that maintainers need to scan through to find issues.

So here's my suggestion: can we use context variables and a custom log formatter that reads a wallet_id context var to retrieve the wallet id when it's available?

shaangill025 commented:
@dbluhm Based on your feedback, I have reworked the logging mechanism.

I added `configure_per_tenant` function to [LoggingConfigurator](https://github.com/hyperledger/aries-cloudagent-python/blob/f9b945b5f3f452c8fd7e92de6584095fb45b0cb4/aries_cloudagent/config/logging.py#L87) which will be called during startup once [here](https://github.com/hyperledger/aries-cloudagent-python/blob/f9b945b5f3f452c8fd7e92de6584095fb45b0cb4/aries_cloudagent/config/util.py#L26). Also added [get_adapted_logger_inst](https://github.com/hyperledger/aries-cloudagent-python/blob/f9b945b5f3f452c8fd7e92de6584095fb45b0cb4/aries_cloudagent/config/logging.py#L576) utility function in `aries_cloudagent/config/logging.py` to improve code readability and reduce duplication, this will be called upon from different classes to get an adapted logger instance [formatted with wallet_id], if applicable.

shaangill025 commented:
> What throws me off with this implementation is there is now a "special" way the get a "special" logger, rather than one consistent way to get a logger. A developer needs to be conscious of which logger they are using in a given context in order to have the desired behavior.

@WadeBarnes The reworked implementation adds a conditional to the existing app configuration to configure logger [here](https://github.com/hyperledger/aries-cloudagent-python/blob/d678ad45d294040115d844c39527345914e25e1a/aries_cloudagent/config/util.py#L12):
- If `--log-file` is passed then `configure_per_tenant` is called. The log handlers will then be added [i.e configuration applied] to all `loggers` across that agent instance.  There are no `special` loggers in case `log.file` setting is involved as this will be the only type of logger.

> It would be better if the implementation was more compatible with the built in logging functionality so it would be completely transparent to the developer. So the implementation may consist of a custom log handler and a custom log formatter. I believe this is inline with @dbluhm's suggestion of a custom log formatter
- Re `implementation may consist of a custom log handler and a custom log formatter`, that is implemented in [`configure_per_tenant`](https://github.com/hyperledger/aries-cloudagent-python/blob/d678ad45d294040115d844c39527345914e25e1a/aries_cloudagent/config/logging.py#L125). This function is called once during startup to apply the configuration to all loggers. But the thing is that each `Logger` instance declared on different files also needs to be adapted with a contextual `wallet_id` identifier from profile settings. And this is done by calling [get_adapted_logger_inst](https://github.com/hyperledger/aries-cloudagent-python/blob/d678ad45d294040115d844c39527345914e25e1a/aries_cloudagent/config/logging.py#L582), you will see calls to this function across most files in this PR.

Does this address your concerns?

shaangill025 commented:
### Performance Demo Metrics

**This PR**

|  config  |  avg time/credential |  total time  |
|:-:|:-:|:-:|
|  --log-file acapy.log --log-handler-config "d;7;1" --log-json-fmt --multitenant --log-level warning |  0.80s  |  240.79s  |
|  --log-file acapy.log --log-handler-config "d;7;1" --log-json-fmt --multitenant --log-level warning  |  0.74s |  221.11s  |
|   |   |   |

**Main**

|  config  |  avg time/credential |  total time  |
|:-:|:-:|:-:|
|  --multitenant --log-level warning  |  0.74s  |  223.17s |
|  --multitenant --log-level warning  |  1.01s |  304.08s  |
|   |   |   |

WadeBarnes commented:
I'm not super familiar with extending logging in python, but per-tenant logging, on the surface, to me, feels like an extension to the built in file log handler that collects the additional context and writes to individual files (one for each tenant), or a custom log formatter that collects and writes the additional context to a fie logger configured to write to a dynamic set of files (one for each tenant).

shaangill025 commented:
@WadeBarnes @esune  This is ready for review, all changes requested should be incorporated now [incl `DictConfig` support]. The `Logging.md` has been updated and can be of help in testing this out.

swcurran commented:
Let’s get this reviewed and merged in the next few days. Causing grief in debugging in multi-tenant worlds.

Thanks!

WadeBarnes commented:
@esune, @loneil, @shaangill025, @swcurran, I've created a ore release image for this PR if you'd like to test it out with a traction instance; https://github.com/hyperledger/aries-cloudagent-python/pkgs/container/aries-cloudagent-python/134985613?tag=py3.9-0.11.0-pr2425-pre0

esune commented on aries_cloudagent/config/logging.py:
```diff
@@ -1,30 +1,57 @@
+LOG_FORMAT_FILE_ALIAS_PATTERN = (
+    "%(asctime)s %(wallet_id)s %(levelname)s %(pathname)s:%(lineno)d %(message)s"
+)
+LOG_FORMAT_STREAM_PATTERN = (
+    "%(asctime)s %(levelname)s %(pathname)s:%(lineno)d %(message)s"
+)
```
This does not seem to have any references in the new code?

esune commented on aries_cloudagent/config/logging.py:
```diff
@@ -1,30 +1,57 @@
+DEFAULT_PER_TENANT_LOGGING_CONFIG_PATH = (
+    "./aries_cloudagent/config/default_per_tenant_logging_config.yml"
+)
+LOG_FORMAT_FILE_ALIAS_PATTERN = (
+    "%(asctime)s %(wallet_id)s %(levelname)s %(pathname)s:%(lineno)d %(message)s"
+)
```
This looks like the same pattern/variable that is held in `default_per_tenant_logging_config.ini` - any reason to have the setting it repeated here?

esune reviewed (commented):
This is looking a lot better than the first pass 🙂 
I added a couple comments/questions, but other than that it looks good to me. If @dbluhm and/or @usingtechnology want to give it a look as they are more frequently involved in code changes for ACA-Py it would be great.

WadeBarnes commented:
@esune, @loneil, @shaangill025, @swcurran, A new pre release image for this PR has been created; https://github.com/hyperledger/aries-cloudagent-python/pkgs/container/aries-cloudagent-python/137906897?tag=py3.9-0.11.0-pr2425-pre1

WadeBarnes commented:
Pre image was rebuilt; https://github.com/hyperledger/aries-cloudagent-python/pkgs/container/aries-cloudagent-python/137970106?tag=py3.9-0.11.0-pr2425-pre2

shaangill025 commented:
Closing this PR and opening #2550 due to an issue with legacy (removed) code being in the pre-release images.
